File: 2020/07/2020-07.jl


test_data = """
light red bags contain 1 bright white bag, 2 muted yellow bags.
dark orange bags contain 3 bright white bags, 4 muted yellow bags.
bright white bags contain 1 shiny gold bag.
muted yellow bags contain 2 shiny gold bags, 9 faded blue bags.
shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags.
dark olive bags contain 3 faded blue bags, 4 dotted black bags.
vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.
faded blue bags contain no other bags.
dotted black bags contain no other bags."""

function parse_rule(line)
    f = split(line)
    key = "$(f[1]) $(f[2])"
    kids = Vector{Tuple{Int,String}}()
    @assert f[3] == "bags" && f[4] == "contain"
    if f[5] != "no"
        rs = f[5:end]
        while length(rs) > 0
            @assert length(rs) >= 4
            num = parse(Int, rs[1])
            col = "$(rs[2]) $(rs[3])"
            push!(kids, (num, col))
            rs = rs[5:end]
        end
    end
    (key, kids)
end

@assert parse_rule("vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.") == ("vibrant plum", [(5, "faded blue"), (6, "dotted black")])
@assert parse_rule("faded blue bags contain no other bags.") == ("faded blue", [])

read_data(inp) = [parse_rule(line) for line in split(strip(inp), "\n")]

td = read_data(test_data)

function part1(data)
    g = Dict{String,Set{String}}()
    for (parentcolor,kids) in data
        for (num,kidcolor) in kids
            push!(get!(g, kidcolor, Set()), parentcolor)
        end
    end
    parents = Set{String}()
    kids = ["shiny gold"]
    while length(kids) > 0
        kids2 = []
        for kidcolor in kids
            ps = get(g, kidcolor, [])
            append!(kids2, ps)
            union!(parents, ps)
        end
        kids = kids2
    end
    parents
end

@assert length(part1(td)) == 4

data = read_data(read("input.txt", String))
println("PART1: ", length(part1(data)))

function nbags(g, col)
    kids = g[col]
    length(kids) == 0 ? 1 : 1 + sum(n * nbags(g, c) for (n,c) in kids)
end

function part2(data)
    g = Dict((k => v) for (k,v) in data)
    nbags(g, "shiny gold") - 1
end

@assert part2(td) == 32

test_data2 = """
shiny gold bags contain 2 dark red bags.
dark red bags contain 2 dark orange bags.
dark orange bags contain 2 dark yellow bags.
dark yellow bags contain 2 dark green bags.
dark green bags contain 2 dark blue bags.
dark blue bags contain 2 dark violet bags.
dark violet bags contain no other bags."""

@assert part2(read_data(test_data2)) == 126

println("PART2: ", part2(data))

File: 2020/07/2020-07.exs
defmodule Advent07 do
  def test_data do
    ~S"""
    light red bags contain 1 bright white bag, 2 muted yellow bags.
    dark orange bags contain 3 bright white bags, 4 muted yellow bags.
    bright white bags contain 1 shiny gold bag.
    muted yellow bags contain 2 shiny gold bags, 9 faded blue bags.
    shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags.
    dark olive bags contain 3 faded blue bags, 4 dotted black bags.
    vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.
    faded blue bags contain no other bags.
    dotted black bags contain no other bags.
    """
  end

  def parse_rule(s) do
    rr = Regex.scan(~r/(\d+ )?(\w+ \w+) bag/, s, capture: :all_but_first)
    ["", main_color] = hd rr
    kids = case tl rr do
      [["", "no other"]] -> []
      ls -> for [num_str, color_name] <- ls, do:
        {color_name, String.to_integer(String.trim(num_str))}
    end
    {main_color, kids}
  end

  def read_data(inp) do
    String.trim(inp)
    |> String.split("\n")
    |> Enum.map(&parse_rule(&1))
    |> Enum.into(%{})
  end

  def ibg(rr, c) do
    fs = Map.get(rr, c, [])
    [c | Enum.flat_map(fs, &ibg(rr, &1))]
  end

  def part1(rules) do
    rr = for {c,kids} <- rules, {kc,_n} <- kids, reduce: %{} do
      acc -> Map.update(acc, kc, [c], &[c | &1])
    end
    length(Enum.uniq(ibg(rr, "shiny gold"))) - 1
  end

  def nbg(rules, mult, col) do
    mult +
      Enum.sum(for {c,n} <- rules[col], do: nbg(rules, n*mult, c))
  end

  def part2(rules), do: nbg(rules, 1, "shiny gold") - 1
end

ExUnit.start
defmodule Advent.Test do
  use ExUnit.Case
  import Advent07

  test "test" do
    td = read_data(test_data())
    data = read_data(File.read!("input.txt"))
    assert part1(td) == 4
    IO.puts "PART1: #{part1(data)}"
    IO.puts "PART2: #{part2(data)}"
  end
end
